### PR TITLE
Update DistConv support for Cosmoflow and U-Net models

### PR DESCRIPTION
**Pull request by bvanessen**
Updated the HDF5 legacy data reader and U-Net model to use the new
input layer API.  Added a routine to fetch data fields for label reconstruction.

### CODE REVIEW

**Comment by bvanessen** [comment]
It is currently failing on pascal with `CUDNN_LOGINFO_DBG=1 CUDNN_LOGWARN_DBG=1 CUDNN_LOGERR_DBG=1 CUDNN_LOGDEST_DBG=cudnn_log_%i.txt python3 unet3d.py --nodes 1 --procs-per-node 2 --partition-level 2 --depth-groups 2`

**Comment by bvanessen** [comment]
@fiedorowicz1 This PR now has partial support for multiple samples per rank in the Cosmoflow application.

File: applications/physics/cosmology/cosmoflow/cosmoflow_network_architectures.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
+import lbann
+import lbann.models
+import lbann.modules as lm
+
```
Comment (fiedorowicz1) [inline applications/physics/cosmology/cosmoflow/cosmoflow_network_architectures.py:4]: This file needs a numpy import: `import numpy as np`

File: include/lbann/utils/exception.hpp [hunk 2]
```diff
@@ -135,7 +136,7 @@ using lbann_exception = exception;
 template <typename... Args>
 std::string build_string(Args&&... args) {
   std::ostringstream oss;
-  int dummy[] = { (oss << args, 0)... };
+  int dummy[] = { (oss << std::setfill('0') << std::setw(7) << args, 0)... };
```
Comment (benson31) [inline include/lbann/utils/exception.hpp:139]:
```suggestion
  int dummy[] = { (oss << args, 0)... };
```

File: include/lbann/data_readers/data_reader_hdf5_legacy.hpp [hunk 3]
```diff
@@ -131,11 +131,13 @@ class hdf5_reader : public generic_data_reader {
                            TensorDataType *sample);
   void read_hdf5_sample(int data_id, TensorDataType *sample, TensorDataType *labels);
   //void set_defaults() override;
+  void load_sample(conduit::Node& node,
+                   int data_id);
```
Comment (benson31) [inline include/lbann/data_readers/data_reader_hdf5_legacy.hpp:135]:
```suggestion
  void load_sample(conduit::Node& node, int data_id);
```

File: include/lbann/layers/layer.hpp [hunk 4]
```diff
@@ -188,22 +188,79 @@ struct ParallelStrategy {
...
+     << ", " << "H: " << "G" << "/" << "S"
+     << ", " << "W: " << "G" << "/" << "S"
+     << ", " << "F: " << "G" << "/" << "S"
+     << ", " << "R:"
+     << ", " << "T:"
+     << ", " << "%:"
+     << ", " << "e:"
      << "}";
   return os;
 }
```
Comment (benson31) [inline include/lbann/layers/layer.hpp:266]:
```suggestion
static std::ostream &print_parallel_strategy_header(std::ostream &os) {
  os << "Axis over which DistConv can parallelize:\n"
     << "\tSamples in the mini-batch (N)\n"
     << "\tDepth, Height, and Width (D x H x W)\n"
     << "\tChannel (C)\n"
     << "\tFilters (F)\n"
     << "\tReplications (R): Number of times the layer is replicated (for FC layers right now)\n"
     << "\tBranch number in the subgraph (T)\n"
     << "\tPercentage of parent resources to be allocated to this branch (%)\n"
     << "\tEnable subgraph for the layer (e)\n"
     << "\nFor each of the above dimensions there are two fields:\n"
     <<"\t# Groups (G): refers to how many reduced-order tensors exist with respect to that dimension" << std::endl
     << "\t             e.g. For a kD tensor you would have a stack of G (k-1)D tensors" << std::endl
     << "\t\t[N, C, D, H, W]" << std::endl
     << "\t\t[2, 1, 4, 1, 1] ---" << std::endl
     << "\t\t                  |" << std::endl
     << "\t\t                  V" << std::endl
     << "\t\t  4 Depth groups: [N, C, H, W]" << std::endl
     << "\t\t                  [2, 1, 1, 1]" << std::endl
     << "\t\t                  [2, 1, 1, 1]" << std::endl
     << "\t\t                  [2, 1, 1, 1]" << std::endl
     << "\t\t                  [2, 1, 1, 1]" << std::endl
     << "\t\t[1, 1, 4, 1, 2] ---" << std::endl
     << "\t\t                  |" << std::endl
     << "\t\t                  V" << std::endl
     << "\t\t 2 Sample groups: [C, D, H, W]" << std::endl
     << "\t\t                  [1, 4, 1, 1]" << std::endl
     << "\t\t                  [1, 4, 1, 1]" << std::endl
     << "\n\tSplit per Dimension (S): Number of groups the dimension is split over (i.e. split K times) (aka H2 split shape) (must divide groups evenly).\n"
     << std::endl;


  os << "Reporting order for the parallel strategy" << std::endl;
  os << "{"  << "N: " << "G" << "/" << "S"
     << ", " << "C: " << "G" << "/" << "S"
     << ", " << "D: " << "G" << "/" << "S"
     << ", " << "H: " << "G" << "/" << "S"
     << ", " << "W: " << "G" << "/" << "S"
     << ", " << "F: " << "G" << "/" << "S"
     << ", " << "R:"
     << ", " << "T:"
     << ", " << "%:"
     << ", " << "e:"
     << "}";
  return os;
}
```
I'm not sure we actually want to print this. It takes a lot of terminal/log file real estate for something that, frankly, is neither interesting nor compilcated.

If we decide to proceed anyway with this, I'd recommend cleaning it up (get rid of the extraneous `<<` tokens and `std::endl`s) and perhaps only printing it in debug builds.

File: python/lbann/contrib/args.py [hunk 5]
```diff
@@ -77,9 +77,11 @@ def get_distconv_environment(parallel_io=False, num_io_partitions=1):
...
     return {
         'DISTCONV_WS_CAPACITY_FACTOR': 0.8,
-        'DISTCONV_OVERLAP_HALO_EXCHANGE': 1,
-        'LBANN_DISTCONV_HALO_EXCHANGE': 'HYBRID',
-        'LBANN_DISTCONV_TENSOR_SHUFFLER': 'HYBRID',
+#        'DISTCONV_OVERLAP_HALO_EXCHANGE': 1,
+        'LBANN_DISTCONV_HALO_EXCHANGE': 'AL',
+        'LBANN_DISTCONV_TENSOR_SHUFFLER': 'AL',
+#        'LBANN_DISTCONV_HALO_EXCHANGE': 'HYBRID',
+#        'LBANN_DISTCONV_TENSOR_SHUFFLER': 'HYBRID',
```
Comment (benson31) [inline python/lbann/contrib/args.py:84]:
```suggestion
        'LBANN_DISTCONV_HALO_EXCHANGE': 'AL',
        'LBANN_DISTCONV_TENSOR_SHUFFLER': 'AL',
```

File: src/models/model.cpp [hunk 6]
```diff
@@ -2667,8 +2667,11 @@ void model::setup_distconv()
     }
   }
   if (m_comm->am_world_master()) {
-    std::cout << "Distconv-enabled layers: " << dc_enabled.str() << std::endl;
-    std::cout << "Distconv-disabled layers: " << dc_disabled.str() << std::endl;
+    std::cout << "\nDistconv-enabled layers:\n\t" << dc_enabled.str() << std::endl << std::endl;
+    std::cout << "Distconv-disabled layers:\n\t" << dc_disabled.str() << std::endl;
+    std::stringstream ss;
+    print_parallel_strategy_header(ss);
+    std::cout << "\nParallel Strategy description - " << ss.str() << std::endl;
```
Comment (benson31) [inline src/models/model.cpp:2674]:
```suggestion
    std::cout << "\nDistconv-enabled layers:\n\t" << dc_enabled.str() << "\n\n"
              << "Distconv-disabled layers:\n\t" << dc_disabled.str() << "\n\n"
              << "Parallel Strategy description - ";
    print_parallel_strategy_header(std::cout);
    endl(std::cout);
```

**Review by benson31 — CHANGES_REQUESTED** [review]
Seems reasonable, but let's discuss and address the excess of print statements this adds.

**Review by benson31 — APPROVED** [review]

**Pull request merged**